Sources/Regex/CharacterClass.swift

```diff
@@ -1,29 +1,94 @@
-public enum CharacterClass {
-  // Any character
-  case any
-  // Character.isDigit
-  case digit
-  // Character.isHexDigit
-  case hexDigit
-  // Character.isWhitespace
-  case whitespace
-  // Character.isLetter or Character.isDigit or Character == "_"
-  case word
+public struct CharacterClass: Hashable {
+  /// The actual character class to match.
+  var cc: Representation
+  
+  /// The level (character or Unicode scalar) at which to match.
+  var matchLevel: MatchLevel
+  
+  public enum Representation {
+    // Any character
+    case any
+    // Character.isDigit
+    case digit
+    // Character.isHexDigit
+    case hexDigit
+    // Character.isWhitespace
+    case whitespace
+    // Character.isLetter or Character.isDigit or Character == "_"
+    case word
+  }
+  
+  public enum MatchLevel {
+    /// Match at the extended grapheme cluster level.
+    case character
+    /// Match at the Unicode scalar level.
+    case unicodeScalar
+  }
+
+  /// Returns the end of the match of this character class in `str`, if
+  /// it matches.
+  public func matches(in str: String, at i: String.Index) -> String.Index? {
+    switch matchLevel {
+    case .character:
+      let c = str[i]
+      let next = str.index(after: i)
+      switch cc {
+      case .any: return next
+      case .digit: return c.isNumber ? next : nil
+      case .hexDigit: return c.isHexDigit ? next : nil
+      case .whitespace: return c.isWhitespace ? next : nil
+      case .word: return c.isLetter || c.isNumber || c == "_"
+        ? next : nil
+      }
+    case .unicodeScalar:
+      let c = str.unicodeScalars[i]
+      let next = str.unicodeScalars.index(after: i)
+      switch cc {
+      case .any: return next
+      case .digit: return c.properties.numericType != nil ? next : nil
+      case .hexDigit: return Character(c).isHexDigit ? next : nil
+      case .whitespace: return c.properties.isWhitespace ? next : nil
+      case .word: return c.properties.isAlphabetic || c == "_" ? next : nil
+      }
+    }
+  }
+}
 
-  public func matches(_ c: Character) -> Bool {
-    switch self {
-    case .any: return true
-    case .digit: return c.isNumber
-    case .hexDigit: return c.isHexDigit
-    case .whitespace: return c.isWhitespace
-    case .word: return c.isLetter || c.isNumber || c == "_"
+extension CharacterClass {
+  public static var any: CharacterClass {
```

Comment:
How will the user create a character class with a different match level?

Reply:
That's a great question — tbh I'm not sure about the utility of distinguishing between capturing substrings, characters, and unicode scalars. If every capture is always a substring, we benefit in terms of simplicity (for things like this) and also providing ranges, since substrings double as a range pointing into the source input.